File: CS/MyLookUpEdit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpf.Core.Native;
using DevExpress.Xpf.Grid;
using System.Windows;
using DevExpress.Data.Async.Helpers;
using DevExpress.Xpf.Editors;
using DevExpress.Xpf.Utils;
using DevExpress.Data;
using System.Reflection;
using System.Windows.Data;
using System.Windows.Controls;
using DevExpress.Xpf.Core;
using DevExpress.Xpf.Editors.Native;
using DevExpress.Xpf.Editors.Validation.Native;
using DevExpress.Xpf.Editors.EditStrategy;
using DevExpress.Xpf.Editors.Popups;

namespace DXGridSample {
    public class MyLookUpEdit : PopupBaseEdit {
        public static readonly DependencyProperty ItemsSourceProperty =
            DependencyPropertyManager.Register("ItemsSource", typeof(object), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(null));
        public static readonly DependencyProperty DisplayMemberProperty =
            DependencyPropertyManager.Register("DisplayMember", typeof(string), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(string.Empty));
        public static readonly DependencyProperty SelectedValueProperty =
            DependencyPropertyManager.Register("SelectedValue", typeof(object), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(null, OnSelectedValuePropertyChanged));
        public static readonly DependencyProperty ValueMemberProperty =
            DependencyPropertyManager.Register("ValueMember", typeof(string), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(null));
        public static readonly DependencyProperty GridControlProperty =
            DependencyPropertyManager.Register("GridControl", typeof(GridControl), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(null, OnGridControlPropertyChanged));
        public static readonly DependencyProperty FocusedRowHandleProperty =
            DependencyProperty.Register("FocusedRowHandle", typeof(int), typeof(MyLookUpEdit), new FrameworkPropertyMetadata(GridControl.InvalidRowHandle, OnFocusedRowHandlePropertyChanged));
        public static readonly DependencyProperty IncrementalFilteringProperty =
            DependencyProperty.Register("IncrementalFiltering", typeof(bool), typeof(MyLookUpEdit), new UIPropertyMetadata(true));
        public static readonly DependencyProperty GridSearchStringProperty =
            DependencyProperty.Register("GridSearchString", typeof(string), typeof(MyLookUpEdit), new UIPropertyMetadata(string.Empty));

        static MyLookUpEdit() {
            AllowRecreatePopupContentProperty.OverrideMetadata(typeof(MyLookUpEdit), new FrameworkPropertyMetadata(false));
            EditValueProperty.OverrideMetadata(typeof(MyLookUpEdit), new FrameworkPropertyMetadata(null, OnEditValuePropertyChanged));
        }

        #region Properties
        public GridControl GridControl {
            get { return (GridControl)GetValue(GridControlProperty); }
            set { SetValue(GridControlProperty, value); }
        }
        public object ItemsSource {
            get { return (object)GetValue(ItemsSourceProperty); }
            set { SetValue(ItemsSourceProperty, value); }
        }
        public string DisplayMember {
            get { return (string)GetValue(DisplayMemberProperty); }
            set { SetValue(DisplayMemberProperty, value); }
        }
        public string ValueMember {
            get { return (string)GetValue(ValueMemberProperty); }
            set { SetValue(ValueMemberProperty, value); }
        }
        public object SelectedValue {
            get { return (object)GetValue(SelectedValueProperty); }
            set { SetValue(SelectedValueProperty, value); }
        }
        public int FocusedRowHandle {
            get { return (int)GetValue(FocusedRowHandleProperty); }
            set { SetValue(FocusedRowHandleProperty, value); }
        }
        public bool IncrementalFiltering {
            get { return (bool)GetValue(IncrementalFilteringProperty); }
            set { SetValue(IncrementalFilteringProperty, value); }
        }
        public string GridSearchString {
            get { return (string)GetValue(GridSearchStringProperty); }
            set { SetValue(GridSearchStringProperty, value); }
        }
        public DataController Controller { get { return GridControl.DataController; } }
        MethodInfo processForegroundMethodInfo;
        MethodInfo ProcessForegroundMethodInfo {
            get {
                if (processForegroundMethodInfo == null)
                    processForegroundMethodInfo = EditStrategy.GetType().GetMethod("ProcessForeground", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
                return processForegroundMethodInfo;
            }
        }
        #endregion

        #region PropertyChangedCallbacks
        static void OnEditValuePropertyChanged(DependencyObject o, DependencyPropertyChangedEventArgs e) {
            ((MyLookUpEdit)o).OnEditValueChanged(e);
        }
        static void OnSelectedValuePropertyChanged(DependencyObject o, DependencyPropertyChangedEventArgs e) {
            ((MyLookUpEdit)o).OnSelectedValueChanged(e);
        }
        static void OnGridControlPropertyChanged(DependencyObject o, DependencyPropertyChangedEventArgs e) {
            ((MyLookUpEdit)o).OnGridControlChanged(e);
        }
        static void OnFocusedRowHandlePropertyChanged(DependencyObject o, DependencyPropertyChangedEventArgs e) {
            ((MyLookUpEdit)o).OnFocusedRowHandleChanged(e);
        }
        #endregion

        protected override EditStrategyBase CreateEditStrategy() {
            return new MyEditStrategy(this);
        }

        protected virtual void OnEditValueChanged(DependencyPropertyChangedEventArgs e) {
            if (string.IsNullOrEmpty(ValueMember)) {
                SelectedValue = EditValue;
                return;
            }
            UpdateFocusedRowHandle();
        }
        protected virtual void OnSelectedValueChanged(DependencyPropertyChangedEventArgs e) {
            if (string.IsNullOrEmpty(ValueMember)) {
                EditValue = SelectedValue;
                return;
            }
            EditValue = GetPropertyValue(e.NewValue, ValueMember);
        }
        protected virtual void OnFocusedRowHandleChanged(DependencyPropertyChangedEventArgs e) {
            SelectedValue = GridControl.GetRow(FocusedRowHandle);
        }
        protected virtual void OnGridControlChanged(DependencyPropertyChangedEventArgs e) {
            SubscribeOnEvents();
            GridControl.SetBinding(GridControl.ItemsSourceProperty, new Binding("ItemsSource") { Source = this });
        }
        void OnRowHandleFound(object obj) {
            if ((int)obj < 0)
                return;
            object selectedValue = GridControl.DataController.GetRow((int)obj, OnRowObjectFound);
            if (ValueIsCorrect(selectedValue) && !(selectedValue is NotLoadedObject)) {
                SelectedValue = selectedValue;
            }
        }
        bool ValueIsCorrect(object value) {
            if (value == null)
                return false;
            return value is ReadonlyThreadSafeProxyForObjectFromAnotherThread ? ((ReadonlyThreadSafeProxyForObjectFromAnotherThread)value).OriginalRow != null : true;
        }
        void OnRowObjectFound(object value) {
            SelectedValue = value;
        }

        protected override void OnEditBoxTextChanged(object sender, TextChangedEventArgs e) {
            GridSearchString = this.EditBox.Text;
            if (GridControl.View != null)
                GridControl.View.SearchString = GridSearchString;
            var controllerRow = Controller.FindRowByValue(DisplayMember, GridSearchString, n => {
                if ((int)n > -1)
                    Controller.EnsureRowLoaded((int)n, new OperationCompleted(nn => { SelectedValue = nn; base.OnEditBoxTextChanged(sender, e); }));
            });
            if (controllerRow >= 0) {
                SelectedValue = Controller.GetRow(controllerRow);
                base.OnEditBoxTextChanged(sender, e);
            }
        }
        protected override PopupSettings CreatePopupSettings() {
            return new MyPopupSettings(this);
        }
        void SetOwnerEdit(DependencyObject element, BaseEdit value) {
            if (element == null)
                throw new ArgumentNullException("element");
            element.SetValue(OwnerEditPropertyKey, value);
        }
        protected override string GetDisplayText(object editValue, bool applyFormatting) {
            if (SelectedValue == null)
                return null;
            if (string.IsNullOrEmpty(DisplayMember))
                return SelectedValue.ToString();
            var value = GetPropertyValue(SelectedValue, DisplayMember);
            if (value != null)
                return value.ToString();
            return base.GetDisplayText(editValue, applyFormatting);
        }

        void SubscribeOnEvents() {
            GridControl.PreviewMouseUp -= OnGridControlPreviewMouseDown;
            GridControl.PreviewKeyDown -= OnGridControlPreviewKeyDown;
            GridControl.Loaded -= OnGridControlLoaded;
            GridControl.ItemsSourceChanged -= OnGridControlItemsSourceChanged;
            GridControl.EndSorting -= OnGridControlEndSorting;
            GridControl.AsyncOperationCompleted -= OnGridControlAsyncOperationCompleted;

            GridControl.PreviewMouseUp += OnGridControlPreviewMouseDown;
            GridControl.PreviewKeyDown += OnGridControlPreviewKeyDown;
            GridControl.Loaded += OnGridControlLoaded;
            GridControl.ItemsSourceChanged += OnGridControlItemsSourceChanged;
            GridControl.EndSorting += OnGridControlEndSorting;
            GridControl.AsyncOperationCompleted += OnGridControlAsyncOperationCompleted;
        }

        void UpdateFocusedRowHandle() {
            if (!string.IsNullOrEmpty(ValueMember)) {
                int i = Controller.FindRowByValue(ValueMember, EditValue, OnRowHandleFound);
                UpdateFocusedRowHandleCore(i);
            }
            else {
                int i = Controller.FindRowByRowValue(EditValue);
                UpdateFocusedRowHandleCore(i);
            }
        }
        void UpdateFocusedRowHandleCore(int i) {
            if (i >= 0 && FocusedRowHandle != i) {
                int rowHandle = GridControl.GetRowHandleByVisibleIndex(Controller.GetVisibleIndex(i));
                FocusedRowHandle = rowHandle;
            }
        }

        #region Grid Event handlers
        protected virtual void OnGridControlPreviewMouseDown(object sender, System.Windows.Input.MouseButtonEventArgs e) {
            if (((TableView)GridControl.View).GetRowElementByMouseEventArgs(e) != null) {
                SelectedValue = GridControl.View.FocusedRow;
                ClosePopup();
            }
        }
        protected virtual void OnGridControlPreviewKeyDown(object sender, System.Windows.Input.KeyEventArgs e) {
            if (e.Key == System.Windows.Input.Key.Enter) {
                SelectedValue = GridControl.View.FocusedRow;
                ClosePopup();
            }
        }
        protected virtual void OnGridControlLoaded(object sender, RoutedEventArgs e) {
            GridControl.View.SearchString = GridSearchString;
            GridControl.View.SetBinding(TableView.SearchPanelAllowFilterProperty, new Binding("IncrementalFiltering") { Source = this });
            Dispatcher.BeginInvoke((Action)GridControl.InvalidateMeasure);
        }
        protected virtual void OnGridControlItemsSourceChanged(object sender, ItemsSourceChangedEventArgs e) {
            UpdateFocusedRowHandle();
        }
        protected virtual void OnGridControlEndSorting(object sender, RoutedEventArgs e) {
            UpdateFocusedRowHandle();
        }
        void OnGridControlAsyncOperationCompleted(object sender, RoutedEventArgs e) {
            UpdateFocusedRowHandle();
        }
        #endregion

        public virtual object GetPropertyValue(object obj, string property) {
            if (obj == null)
                return null;
            object originalRow = obj;
            if (originalRow is ReadonlyThreadSafeProxyForObjectFromAnotherThread)
                originalRow = ((ReadonlyThreadSafeProxyForObjectFromAnotherThread)originalRow).OriginalRow;
            var descr = originalRow.GetType().GetProperty(property);
            if (descr == null)
                return null;
            return descr.GetValue(originalRow, null);
        }
        public virtual object GetEditValueByDisplayValue(object displayValue) {
            if (string.IsNullOrEmpty(DisplayMember))
                if (string.IsNullOrEmpty(ValueMember))
                    return displayValue;
                else
                    GetPropertyValue(displayValue, ValueMember);
            int i = Controller.FindRowByValue(DisplayMember, displayValue, n => { });
            object selectedValue = GridControl.DataController.GetRow(i, value => { });
            if (string.IsNullOrEmpty(ValueMember))
                return selectedValue;
            return GetPropertyValue(selectedValue, ValueMember);
        }
        public IPopupContentOwner PublicPopupContentOwner { get { return PopupContentOwner; } }
    }
    public class MyEditStrategy : TextEditStrategy {
        MyLookUpEdit LookUp;
        public override void EditValueChanged(object oldValue, object newValue) {
            base.EditValueChanged(oldValue, newValue);
        }
        public MyEditStrategy(MyLookUpEdit edit)
            : base(edit) {
            LookUp = edit;
        }
        protected override object GetEditValueInternal() {
            string containersValue = this.EditBox.Text;
            if (string.IsNullOrEmpty(LookUp.DisplayMember) && string.IsNullOrEmpty(LookUp.ValueMember))
                return containersValue;
            var a = LookUp.GetEditValueByDisplayValue(containersValue);
            return a;
        }
    }
    public class MyPopupSettings : PopupSettings {
        MyLookUpEdit LookUp { get { return (MyLookUpEdit)OwnerEdit; } }
        public MyPopupSettings(MyLookUpEdit editor) : base(editor) { }
        protected override FrameworkElement GetPopupContent() {
            bool allowRecreatePopupContent = OwnerEdit.AllowRecreatePopupContent;
            NonLogicalDecorator popupChild = new NonLogicalDecorator();
            PopupContentControl popupContent = null;
            if (!allowRecreatePopupContent)
                popupContent = (PopupContentControl)LookUp.PublicPopupContentOwner.Child;
            if (popupContent == null) {
                popupContent = new PopupContentControl();
                FocusHelper.SetFocusable(popupContent, false);
            }
            popupContent.Editor = OwnerEdit;
            popupChild.Child = popupContent;
            LookUp.PublicPopupContentOwner.Child = popupContent;
            popupContent.Tag = OwnerEdit;
            popupContent.Content = LookUp.GridControl;
            return popupChild;
        }
    }
}

File: VB/MyLookUpEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpf.Core.Native
Imports DevExpress.Xpf.Grid
Imports System.Windows
Imports DevExpress.Data.Async.Helpers
Imports DevExpress.Xpf.Editors
Imports DevExpress.Xpf.Utils
Imports DevExpress.Data
Imports System.Reflection
Imports System.Windows.Data
Imports System.Windows.Controls
Imports DevExpress.Xpf.Core
Imports DevExpress.Xpf.Editors.Native
Imports DevExpress.Xpf.Editors.Validation.Native
Imports DevExpress.Xpf.Editors.EditStrategy
Imports DevExpress.Xpf.Editors.Popups

Namespace DXGridSample
    Public Class MyLookUpEdit
        Inherits PopupBaseEdit

        Public Shared ReadOnly ItemsSourceProperty As DependencyProperty = DependencyPropertyManager.Register("ItemsSource", GetType(Object), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(Nothing))
        Public Shared ReadOnly DisplayMemberProperty As DependencyProperty = DependencyPropertyManager.Register("DisplayMember", GetType(String), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(String.Empty))
        Public Shared ReadOnly SelectedValueProperty As DependencyProperty = DependencyPropertyManager.Register("SelectedValue", GetType(Object), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(Nothing, AddressOf OnSelectedValuePropertyChanged))
        Public Shared ReadOnly ValueMemberProperty As DependencyProperty = DependencyPropertyManager.Register("ValueMember", GetType(String), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(Nothing))
        Public Shared ReadOnly GridControlProperty As DependencyProperty = DependencyPropertyManager.Register("GridControl", GetType(GridControl), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(Nothing, AddressOf OnGridControlPropertyChanged))
        Public Shared ReadOnly FocusedRowHandleProperty As DependencyProperty = DependencyProperty.Register("FocusedRowHandle", GetType(Integer), GetType(MyLookUpEdit), New FrameworkPropertyMetadata(GridControl.InvalidRowHandle, AddressOf OnFocusedRowHandlePropertyChanged))
        Public Shared ReadOnly IncrementalFilteringProperty As DependencyProperty = DependencyProperty.Register("IncrementalFiltering", GetType(Boolean), GetType(MyLookUpEdit), New UIPropertyMetadata(True))
        Public Shared ReadOnly GridSearchStringProperty As DependencyProperty = DependencyProperty.Register("GridSearchString", GetType(String), GetType(MyLookUpEdit), New UIPropertyMetadata(String.Empty))

        Shared Sub New()
            AllowRecreatePopupContentProperty.OverrideMetadata(GetType(MyLookUpEdit), New FrameworkPropertyMetadata(False))
            EditValueProperty.OverrideMetadata(GetType(MyLookUpEdit), New FrameworkPropertyMetadata(Nothing, AddressOf OnEditValuePropertyChanged))
        End Sub

        #Region "Properties"
        Public Property GridControl() As GridControl
            Get
                Return CType(GetValue(GridControlProperty), GridControl)
            End Get
            Set(ByVal value As GridControl)
                SetValue(GridControlProperty, value)
            End Set
        End Property
        Public Property ItemsSource() As Object
            Get
                Return DirectCast(GetValue(ItemsSourceProperty), Object)
            End Get
            Set(ByVal value As Object)
                SetValue(ItemsSourceProperty, value)
            End Set
        End Property
        Public Property DisplayMember() As String
            Get
                Return CStr(GetValue(DisplayMemberProperty))
            End Get
            Set(ByVal value As String)
                SetValue(DisplayMemberProperty, value)
            End Set
        End Property
        Public Property ValueMember() As String
            Get
                Return CStr(GetValue(ValueMemberProperty))
            End Get
            Set(ByVal value As String)
                SetValue(ValueMemberProperty, value)
            End Set
        End Property
        Public Property SelectedValue() As Object
            Get
                Return DirectCast(GetValue(SelectedValueProperty), Object)
            End Get
            Set(ByVal value As Object)
                SetValue(SelectedValueProperty, value)
            End Set
        End Property
        Public Property FocusedRowHandle() As Integer
            Get
                Return CInt((GetValue(FocusedRowHandleProperty)))
            End Get
            Set(ByVal value As Integer)
                SetValue(FocusedRowHandleProperty, value)
            End Set
        End Property
        Public Property IncrementalFiltering() As Boolean
            Get
                Return CBool(GetValue(IncrementalFilteringProperty))
            End Get
            Set(ByVal value As Boolean)
                SetValue(IncrementalFilteringProperty, value)
            End Set
        End Property
        Public Property GridSearchString() As String
            Get
                Return CStr(GetValue(GridSearchStringProperty))
            End Get
            Set(ByVal value As String)
                SetValue(GridSearchStringProperty, value)
            End Set
        End Property
        Public ReadOnly Property Controller() As DataController
            Get
                Return GridControl.DataController
            End Get
        End Property

        Private processForegroundMethodInfo_Renamed As MethodInfo
        Private ReadOnly Property ProcessForegroundMethodInfo() As MethodInfo
            Get
                If processForegroundMethodInfo_Renamed Is Nothing Then
                    processForegroundMethodInfo_Renamed = EditStrategy.GetType().GetMethod("ProcessForeground", System.Reflection.BindingFlags.Instance Or System.Reflection.BindingFlags.NonPublic)
                End If
                Return processForegroundMethodInfo_Renamed
            End Get
        End Property
        #End Region

        #Region "PropertyChangedCallbacks"
        Private Shared Sub OnEditValuePropertyChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(o, MyLookUpEdit).OnMyEditValueChanged(e)
        End Sub
        Private Shared Sub OnSelectedValuePropertyChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(o, MyLookUpEdit).OnSelectedValueChanged(e)
        End Sub
        Private Shared Sub OnGridControlPropertyChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(o, MyLookUpEdit).OnGridControlChanged(e)
        End Sub
        Private Shared Sub OnFocusedRowHandlePropertyChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(o, MyLookUpEdit).OnFocusedRowHandleChanged(e)
        End Sub
#End Region

        Protected Overrides Function CreateEditStrategy() As EditStrategyBase
            Return New MyEditStrategy(Me)
        End Function

        Protected Overridable Sub OnMyEditValueChanged(ByVal e As DependencyPropertyChangedEventArgs)
            If String.IsNullOrEmpty(ValueMember) Then
                SelectedValue = EditValue
                Return
            End If
            UpdateFocusedRowHandle()
        End Sub
        Protected Overridable Sub OnSelectedValueChanged(ByVal e As DependencyPropertyChangedEventArgs)
            If String.IsNullOrEmpty(ValueMember) Then
                EditValue = SelectedValue
                Return
            End If
            EditValue = GetPropertyValue(e.NewValue, ValueMember)
        End Sub
        Protected Overridable Sub OnFocusedRowHandleChanged(ByVal e As DependencyPropertyChangedEventArgs)
            SelectedValue = GridControl.GetRow(FocusedRowHandle)
        End Sub
        Protected Overridable Sub OnGridControlChanged(ByVal e As DependencyPropertyChangedEventArgs)
            SubscribeOnEvents()
            GridControl.SetBinding(GridControl.ItemsSourceProperty, New Binding("ItemsSource") With {.Source = Me})
        End Sub
        Private Sub OnRowHandleFound(ByVal obj As Object)
            If DirectCast(obj, Integer) < 0 Then
                Return
            End If

            Dim selectedValue_Renamed As Object = GridControl.DataController.GetRow(DirectCast(obj, Integer), AddressOf OnRowObjectFound)
            If ValueIsCorrect(selectedValue_Renamed) AndAlso Not(TypeOf selectedValue_Renamed Is NotLoadedObject) Then
                SelectedValue = selectedValue_Renamed
            End If
        End Sub
        Private Function ValueIsCorrect(ByVal value As Object) As Boolean
            If value Is Nothing Then
                Return False
            End If
            Return If(TypeOf value Is ReadonlyThreadSafeProxyForObjectFromAnotherThread, DirectCast(value, ReadonlyThreadSafeProxyForObjectFromAnotherThread).OriginalRow IsNot Nothing, True)
        End Function
        Private Sub OnRowObjectFound(ByVal value As Object)
            SelectedValue = value
        End Sub

        Protected Overrides Sub OnEditBoxTextChanged(ByVal sender As Object, ByVal e As TextChangedEventArgs)
            GridSearchString = Me.EditBox.Text
            If GridControl.View IsNot Nothing Then
                GridControl.View.SearchString = GridSearchString
            End If
            Dim controllerRow = Controller.FindRowByValue(DisplayMember, GridSearchString, Sub(n)
                If CInt((n)) > -1 Then
                    Controller.EnsureRowLoaded(CInt((n)), New OperationCompleted(Sub(nn)
                    SelectedValue = nn
                    MyBase.OnEditBoxTextChanged(sender, e)
                    End Sub))
                End If
            End Sub)
            If controllerRow >= 0 Then
                SelectedValue = Controller.GetRow(controllerRow)
                MyBase.OnEditBoxTextChanged(sender, e)
            End If
        End Sub
        Protected Overrides Function CreatePopupSettings() As PopupSettings
            Return New MyPopupSettings(Me)
        End Function
        Private Sub SetOwnerEdit(ByVal element As DependencyObject, ByVal value As BaseEdit)
            If element Is Nothing Then
                Throw New ArgumentNullException("element")
            End If
            element.SetValue(OwnerEditPropertyKey, value)
        End Sub
        Protected Overrides Function GetDisplayText(ByVal editValue As Object, ByVal applyFormatting As Boolean) As String
            If SelectedValue Is Nothing Then
                Return Nothing
            End If
            If String.IsNullOrEmpty(DisplayMember) Then
                Return SelectedValue.ToString()
            End If
            Dim value = GetPropertyValue(SelectedValue, DisplayMember)
            If value IsNot Nothing Then
                Return value.ToString()
            End If
            Return MyBase.GetDisplayText(editValue, applyFormatting)
        End Function

        Private Sub SubscribeOnEvents()
            RemoveHandler GridControl.PreviewMouseUp, AddressOf OnGridControlPreviewMouseDown
            RemoveHandler GridControl.PreviewKeyDown, AddressOf OnGridControlPreviewKeyDown
            RemoveHandler GridControl.Loaded, AddressOf OnGridControlLoaded
            RemoveHandler GridControl.ItemsSourceChanged, AddressOf OnGridControlItemsSourceChanged
            RemoveHandler GridControl.EndSorting, AddressOf OnGridControlEndSorting
            RemoveHandler GridControl.AsyncOperationCompleted, AddressOf OnGridControlAsyncOperationCompleted

            AddHandler GridControl.PreviewMouseUp, AddressOf OnGridControlPreviewMouseDown
            AddHandler GridControl.PreviewKeyDown, AddressOf OnGridControlPreviewKeyDown
            AddHandler GridControl.Loaded, AddressOf OnGridControlLoaded
            AddHandler GridControl.ItemsSourceChanged, AddressOf OnGridControlItemsSourceChanged
            AddHandler GridControl.EndSorting, AddressOf OnGridControlEndSorting
            AddHandler GridControl.AsyncOperationCompleted, AddressOf OnGridControlAsyncOperationCompleted
        End Sub

        Private Sub UpdateFocusedRowHandle()
            If Not String.IsNullOrEmpty(ValueMember) Then
                Dim i As Integer = Controller.FindRowByValue(ValueMember, EditValue, AddressOf OnRowHandleFound)
                UpdateFocusedRowHandleCore(i)
            Else
                Dim i As Integer = Controller.FindRowByRowValue(EditValue)
                UpdateFocusedRowHandleCore(i)
            End If
        End Sub
        Private Sub UpdateFocusedRowHandleCore(ByVal i As Integer)
            If i >= 0 AndAlso FocusedRowHandle <> i Then
                Dim rowHandle As Integer = GridControl.GetRowHandleByVisibleIndex(Controller.GetVisibleIndex(i))
                FocusedRowHandle = rowHandle
            End If
        End Sub

        #Region "Grid Event handlers"
        Protected Overridable Sub OnGridControlPreviewMouseDown(ByVal sender As Object, ByVal e As System.Windows.Input.MouseButtonEventArgs)
            If CType(GridControl.View, TableView).GetRowElementByMouseEventArgs(e) IsNot Nothing Then
                SelectedValue = GridControl.View.FocusedRow
                ClosePopup()
            End If
        End Sub
        Protected Overridable Sub OnGridControlPreviewKeyDown(ByVal sender As Object, ByVal e As System.Windows.Input.KeyEventArgs)
            If e.Key = System.Windows.Input.Key.Enter Then
                SelectedValue = GridControl.View.FocusedRow
                ClosePopup()
            End If
        End Sub
        Protected Overridable Sub OnGridControlLoaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
            GridControl.View.SearchString = GridSearchString
            GridControl.View.SetBinding(TableView.SearchPanelAllowFilterProperty, New Binding("IncrementalFiltering") With {.Source = Me})
            Dispatcher.BeginInvoke(New Action(AddressOf GridControl.InvalidateMeasure))
        End Sub
        Protected Overridable Sub OnGridControlItemsSourceChanged(ByVal sender As Object, ByVal e As ItemsSourceChangedEventArgs)
            UpdateFocusedRowHandle()
        End Sub
        Protected Overridable Sub OnGridControlEndSorting(ByVal sender As Object, ByVal e As RoutedEventArgs)
            UpdateFocusedRowHandle()
        End Sub
        Private Sub OnGridControlAsyncOperationCompleted(ByVal sender As Object, ByVal e As RoutedEventArgs)
            UpdateFocusedRowHandle()
        End Sub
        #End Region

        Public Overridable Function GetPropertyValue(ByVal obj As Object, ByVal [property] As String) As Object
            If obj Is Nothing Then
                Return Nothing
            End If
            Dim originalRow As Object = obj
            If TypeOf originalRow Is ReadonlyThreadSafeProxyForObjectFromAnotherThread Then
                originalRow = DirectCast(originalRow, ReadonlyThreadSafeProxyForObjectFromAnotherThread).OriginalRow
            End If
            Dim descr = originalRow.GetType().GetProperty([property])
            If descr Is Nothing Then
                Return Nothing
            End If
            Return descr.GetValue(originalRow, Nothing)
        End Function
        Public Overridable Function GetEditValueByDisplayValue(ByVal displayValue As Object) As Object
            If String.IsNullOrEmpty(DisplayMember) Then
                If String.IsNullOrEmpty(ValueMember) Then
                    Return displayValue
                Else
                    GetPropertyValue(displayValue, ValueMember)
                End If
            End If
            Dim i As Integer = Controller.FindRowByValue(DisplayMember, displayValue, Sub(n)
            End Sub)

            Dim selectedValue_Renamed As Object = GridControl.DataController.GetRow(i, Sub(value)
            End Sub)
            If String.IsNullOrEmpty(ValueMember) Then
                Return selectedValue_Renamed
            End If
            Return GetPropertyValue(selectedValue_Renamed, ValueMember)
        End Function
        Public ReadOnly Property PublicPopupContentOwner() As IPopupContentOwner
            Get
                Return PopupContentOwner
            End Get
        End Property
    End Class
    Public Class MyEditStrategy
        Inherits TextEditStrategy

        Private LookUp As MyLookUpEdit
        Public Overrides Sub EditValueChanged(ByVal oldValue As Object, ByVal newValue As Object)
            MyBase.EditValueChanged(oldValue, newValue)
        End Sub
        Public Sub New(ByVal edit As MyLookUpEdit)
            MyBase.New(edit)
            LookUp = edit
        End Sub
        Protected Overrides Function GetEditValueInternal() As Object
            Dim containersValue As String = Me.EditBox.Text
            If String.IsNullOrEmpty(LookUp.DisplayMember) AndAlso String.IsNullOrEmpty(LookUp.ValueMember) Then
                Return containersValue
            End If
            Dim a = LookUp.GetEditValueByDisplayValue(containersValue)
            Return a
        End Function
    End Class
    Public Class MyPopupSettings
        Inherits PopupSettings

        Private ReadOnly Property LookUp() As MyLookUpEdit
            Get
                Return CType(OwnerEdit, MyLookUpEdit)
            End Get
        End Property
        Public Sub New(ByVal editor As MyLookUpEdit)
            MyBase.New(editor)
        End Sub
        Protected Overrides Function GetPopupContent() As FrameworkElement
            Dim allowRecreatePopupContent As Boolean = OwnerEdit.AllowRecreatePopupContent
            Dim popupChild As New NonLogicalDecorator()
            Dim popupContent As PopupContentControl = Nothing
            If Not allowRecreatePopupContent Then
                popupContent = CType(LookUp.PublicPopupContentOwner.Child, PopupContentControl)
            End If
            If popupContent Is Nothing Then
                popupContent = New PopupContentControl()
                FocusHelper.SetFocusable(popupContent, False)
            End If
            popupContent.Editor = OwnerEdit
            popupChild.Child = popupContent
            LookUp.PublicPopupContentOwner.Child = popupContent
            popupContent.Tag = OwnerEdit
            popupContent.Content = LookUp.GridControl
            Return popupChild
        End Function
    End Class
End Namespace